# Data Wrangling

**Data wrangling, sometimes referred to as data munging, 
is the process of transforming and mapping data from one "raw" data form into another 
format with the intent of making it more appropriate 
and valuable for a variety of downstream purposes such as analytics.**
*Wikipedia*

## Importing data

In [1]:
import pandas as pd
import numpy as np

In [2]:
personality_data = pd.read_csv('personality_scores.csv',sep=';')

## Looking at the data

In [3]:
personality_data.head()

ID Section 5 of 6 [I am always prepared.]  \
0   0                                 (3, 5)   
1   1                                 (3, 5)   
2   2                                 (3, 5)   
3   3                                 (3, 5)   
4   4                                 (3, 3)   

  Section 5 of 6 [I am easily disturbed.]  \
0                                  (4, 5)   
1                                  (4, 5)   
2                                  (4, 3)   
3                                  (4, 5)   
4                                  (4, 5)   

  Section 5 of 6 [I am exacting (demanding) in my work.]  \
0                                             (3, 5)       
1                                             (3, 5)       
2                                             (3, 3)       
3                                             (3, 3)       
4                                             (3, 3)       

  Section 5 of 6 [I am full of ideas.]  \
0                               (5, 5)   
1                               (5, 5)   
2                               (5, 5)   
3                               (5, 5)   
4                               (5, 3)   

  Section 5 of 6 [I am interested in people.]  \
0                                      (2, 3)   
1                                      (2, 5)   
2                                      (2, 5)   
3                                      (2, 5)   
4                                      (2, 3)   

  Section 5 of 6 [I am not interested in abstract ideas.]  \
0                                             (5, 3)        
1                                             (5, 3)        
2                                             (5, 5)        
3                                             (5, 3)        
4                                             (5, 3)        

  Section 5 of 6 [I am not interested in other people's problems.]  \
0                                             (2, 3)                 
1                                             (2, 5)                 
2                                             (2, 5)                 
3                                             (2, 3)                 
4                                             (2, 3)                 

  Section 5 of 6 [I am not really interested in others.]  \
0                                             (2, 5)       
1                                             (2, 5)       
2                                             (2, 5)       
3                                             (2, 3)       
4                                             (2, 3)       

  Section 5 of 6 [I am quick to understand things.]  ... Unnamed: 60  \
0                                            (5, 5)  ...         NaN   
1                                            (5, 5)  ...         NaN   
2                                            (5, 5)  ...         NaN   
3                                            (5, 3)  ...         NaN   
4                                            (5, 5)  ...         NaN   

  Unnamed: 61 Unnamed: 62 Unnamed: 63 Unnamed: 64 Unnamed: 65 Unnamed: 66  \
0         NaN         NaN         NaN         NaN         NaN         NaN   
1         NaN         NaN         NaN         NaN         NaN         NaN   
2         NaN         NaN         NaN         NaN         NaN         NaN   
3         NaN         NaN         NaN         NaN         NaN         NaN   
4         NaN         NaN         NaN         NaN         NaN         NaN   

  Unnamed: 67 Unnamed: 68 IPIP_HIGH_RISK  
0         NaN         NaN            NaN  
1         NaN         NaN            NaN  
2         NaN         NaN            NaN  
3         NaN         NaN            NaN  
4         NaN         NaN            NaN  

[5 rows x 70 columns]

In [4]:
personality_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1555 entries, 0 to 1554
Data columns (total 70 columns):
ID                                                                          1555 non-null int64
Section 5 of 6 [I am always prepared.]                                      1555 non-null object
Section 5 of 6 [I am easily disturbed.]                                     1555 non-null object
Section 5 of 6 [I am exacting (demanding) in my work.]                      1555 non-null object
Section 5 of 6 [I am full of ideas.]                                        1555 non-null object
Section 5 of 6 [I am interested in people.]                                 1555 non-null object
Section 5 of 6 [I am not interested in abstract ideas.]                     1555 non-null object
Section 5 of 6 [I am not interested in other people's problems.]            1555 non-null object
Section 5 of 6 [I am not really interested in others.]                      1555 non-null object
Section 5 of 6 [I am 

*This is a data set of 1555 entries and 70 features.There are 19 columns with only  null values.*

## Find Duplicate rows based on ID

In [5]:
len(personality_data.ID.unique())
print('There are',len(personality_data.ID.unique()), 'unique IDs')

There are 1555 unique IDs


*The dataset has 1555 ID entries and there are 1555 unique IDs.
This explains that there are no duplicates with reference to the IDs*

## Drop any duplicates that exist

In [6]:
data=personality_data.drop_duplicates(subset=['ID'])

In [7]:
print('After executing the function that drops duplicates the length of the dataset is',len(data))

After executing the function that drops duplicates the length of the dataset is 1555


*This emphasises that there are no duplicates.*

### Comparing the length of the new data frame with old one that may have had duplicates.

*Use an assert statement to check that 
the new data frame is the length of the unique entries of the original data frame.*

*Python's assert statement is a debugging aid that tests a condition.
If the condition is true, it does nothing and your program just continues to execute.
But if the assert condition evaluates to false,
it raises an AssertionError exception with an optional error message.
The assert statement should show that the length of unique values of the original 
data is the same as the length of the  data set were duplicates are dropped.*


In [8]:
old_data = len(personality_data.ID.unique())
new_data = len(data)


In [9]:
assert old_data == new_data

*The unique values of old data set has the same length of 1555
as the new data set. New data set is the old data set without duplicate values.*

### Create new columns containing the total score of each of the personality test subscales

In [10]:
data=data.dropna(axis='columns')

In [11]:
Dict={1:'Extraversion', 2:'Agreeableness', 3:'Conscientiousness',
      4:'Emotional Stability',5:'Intellect'}

In [12]:
data=data.set_index('ID')


In [17]:
list_of_rows=data.values.tolist()

In [33]:
def my_function():
    p=[]
    for l in list_of_rows:
        l=[eval(l[i]) for i in range(len(l))]

        totals = {}
        for ad, x in l:
            if ad not in totals :
                totals[ad] = x
            else :
                totals[ad] += x    
        p.append(totals)
        personality_score=pd.DataFrame(p)
        personality_score.columns = personality_score.columns.to_series().map(Dict)
    return personality_score
my_function() 

Conscientiousness  Emotional Stability  Intellect  Agreeableness  \
0                    48                   36         42             40   
1                    46                   40         42             46   
2                    40                   38         42             40   
3                    38                   40         38             38   
4                    46                   38         36             34   
...                 ...                  ...        ...            ...   
1550                 38                   44         44             48   
1551                 40                   44         34             48   
1552                 40                   46         48             44   
1553                 48                   36         44             38   
1554                 42                   40         30             44   

      Extraversion  
0               30  
1               42  
2               28  
3               30  
4               28  
...            ...  
1550            32  
1551            40  
1552            42  
1553            36  
1554            22  

[1555 rows x 5 columns]

## merging the dataframes

In [38]:
 result = pd.concat([data,my_function() ], axis=1, ignore_index=False)

In [40]:
result

Section 5 of 6 [I am always prepared.]  \
0                                    (3, 5)   
1                                    (3, 5)   
2                                    (3, 5)   
3                                    (3, 5)   
4                                    (3, 3)   
...                                     ...   
1550                                 (3, 5)   
1551                                 (3, 3)   
1552                                 (3, 5)   
1553                                 (3, 5)   
1554                                 (3, 3)   

     Section 5 of 6 [I am easily disturbed.]  \
0                                     (4, 5)   
1                                     (4, 5)   
2                                     (4, 3)   
3                                     (4, 5)   
4                                     (4, 5)   
...                                      ...   
1550                                  (4, 5)   
1551                                  (4, 5)   
1552                                  (4, 3)   
1553                                  (4, 5)   
1554                                  (4, 5)   

     Section 5 of 6 [I am exacting (demanding) in my work.]  \
0                                                (3, 5)       
1                                                (3, 5)       
2                                                (3, 3)       
3                                                (3, 3)       
4                                                (3, 3)       
...                                                 ...       
1550                                             (3, 1)       
1551                                             (3, 5)       
1552                                             (3, 5)       
1553                                             (3, 5)       
1554                                             (3, 1)       

     Section 5 of 6 [I am full of ideas.]  \
0                                  (5, 5)   
1                                  (5, 5)   
2                                  (5, 5)   
3                                  (5, 5)   
4                                  (5, 3)   
...                                   ...   
1550                               (5, 5)   
1551                               (5, 3)   
1552                               (5, 5)   
1553                               (5, 5)   
1554                               (5, 5)   

     Section 5 of 6 [I am interested in people.]  \
0                                         (2, 3)   
1                                         (2, 5)   
2                                         (2, 5)   
3                                         (2, 5)   
4                                         (2, 3)   
...                                          ...   
1550                                      (2, 5)   
1551                                      (2, 5)   
1552                                      (2, 5)   
1553                                      (2, 5)   
1554                                      (2, 5)   

     Section 5 of 6 [I am not interested in abstract ideas.]  \
0                                                (5, 3)        
1                                                (5, 3)        
2                                                (5, 5)        
3                                                (5, 3)        
4                                                (5, 3)        
...                                                 ...        
1550                                             (5, 5)        
1551                                             (5, 3)        
1552                                             (5, 5)        
1553                                             (5, 5)        
1554                                             (5, 3)        

     Section 5 of 6 [I am not interested in other people's problems.]  \
0                                                (2, 3)                 
1                        

### Import the data in departments.csv 

*Merge this data frame with the
personality score data frame, keeping all applicants within the department data frame.*

In [44]:
department_data = pd.read_csv('departments.csv',sep=(';'))

In [45]:
department_data

ID Department
0        0       Data
1        1       Data
2        2       Data
3        3       Data
4        4       Data
...    ...        ...
1550  1550    Web dev
1551  1551    Web dev
1552  1552    Web dev
1553  1553    Web dev
1554  1554    Web dev

[1555 rows x 2 columns]

In [46]:
 merge = pd.concat([result, department_data] ,axis=1, ignore_index=False)

In [47]:
merge

Section 5 of 6 [I am always prepared.]  \
0                                    (3, 5)   
1                                    (3, 5)   
2                                    (3, 5)   
3                                    (3, 5)   
4                                    (3, 3)   
...                                     ...   
1550                                 (3, 5)   
1551                                 (3, 3)   
1552                                 (3, 5)   
1553                                 (3, 5)   
1554                                 (3, 3)   

     Section 5 of 6 [I am easily disturbed.]  \
0                                     (4, 5)   
1                                     (4, 5)   
2                                     (4, 3)   
3                                     (4, 5)   
4                                     (4, 5)   
...                                      ...   
1550                                  (4, 5)   
1551                                  (4, 5)   
1552                                  (4, 3)   
1553                                  (4, 5)   
1554                                  (4, 5)   

     Section 5 of 6 [I am exacting (demanding) in my work.]  \
0                                                (3, 5)       
1                                                (3, 5)       
2                                                (3, 3)       
3                                                (3, 3)       
4                                                (3, 3)       
...                                                 ...       
1550                                             (3, 1)       
1551                                             (3, 5)       
1552                                             (3, 5)       
1553                                             (3, 5)       
1554                                             (3, 1)       

     Section 5 of 6 [I am full of ideas.]  \
0                                  (5, 5)   
1                                  (5, 5)   
2                                  (5, 5)   
3                                  (5, 5)   
4                                  (5, 3)   
...                                   ...   
1550                               (5, 5)   
1551                               (5, 3)   
1552                               (5, 5)   
1553                               (5, 5)   
1554                               (5, 5)   

     Section 5 of 6 [I am interested in people.]  \
0                                         (2, 3)   
1                                         (2, 5)   
2                                         (2, 5)   
3                                         (2, 5)   
4                                         (2, 3)   
...                                          ...   
1550                                      (2, 5)   
1551                                      (2, 5)   
1552                                      (2, 5)   
1553                                      (2, 5)   
1554                                      (2, 5)   

     Section 5 of 6 [I am not interested in abstract ideas.]  \
0                                                (5, 3)        
1                                                (5, 3)        
2                                                (5, 5)        
3                                                (5, 3)        
4                                                (5, 3)        
...                                                 ...        
1550                                             (5, 5)        
1551                                             (5, 3)        
1552                                             (5, 5)        
1553                                             (5, 5)        
1554                                             (5, 3)        

     Section 5 of 6 [I am not interested in other people's problems.]  \
0                                                (2, 3)                 
1                        

### The number of rows in data frames

*Use an assert statement to check that the newly created merged data frame has the same amount of 
rows as the department data frame.*

In [48]:
assert len(merge)==len(department_data)

In [49]:
print('The number of rows in the newly merged data frame is:',len(merge))
print('The number of rows in the department data frame is:',len(department_data))

The number of rows in the newly merged data frame is: 1555
The number of rows in the department data frame is: 1555


*The merged data frame and the department data frame have the same number of rows.*

### The expected number of columns.

*The expected number of columns on a merged DataFrame is 57 columns. 
The reason being that Personality data has 55 columns(Empty columns are dropped)
and Department data has 52 columns. So there are 57 expected columns.*

In [50]:
assert len(merge.columns)==57

In [51]:
print('The merged data frame columns in number are:',len(merge.columns))

The merged data frame columns in number are: 57


### Filter the merged data frame

*Filter the merged data frame so that you get only the applicants
who scored less than 30 on emotional stability, conscientiousness AND agreeableness*

In [52]:
score = merge[(merge['Conscientiousness']<30) & (merge['Agreeableness']<30) & (merge['Emotional Stability'] <30)]

In [53]:
score

Section 5 of 6 [I am always prepared.]  \
881                                  (3, 3)   
1197                                 (3, 5)   

     Section 5 of 6 [I am easily disturbed.]  \
881                                   (4, 1)   
1197                                  (4, 5)   

     Section 5 of 6 [I am exacting (demanding) in my work.]  \
881                                              (3, 1)       
1197                                             (3, 1)       

     Section 5 of 6 [I am full of ideas.]  \
881                                (5, 5)   
1197                               (5, 1)   

     Section 5 of 6 [I am interested in people.]  \
881                                       (2, 1)   
1197                                      (2, 1)   

     Section 5 of 6 [I am not interested in abstract ideas.]  \
881                                              (5, 3)        
1197                                             (5, 3)        

     Section 5 of 6 [I am not interested in other people's problems.]  \
881                                              (2, 5)                 
1197                                             (2, 5)                 

     Section 5 of 6 [I am not really interested in others.]  \
881                                              (2, 3)       
1197                                             (2, 1)       

     Section 5 of 6 [I am quick to understand things.]  \
881                                             (5, 3)   
1197                                            (5, 1)   

     Section 5 of 6 [I am quiet around strangers.]  ...  \
881                                         (1, 5)  ...   
1197                                        (1, 5)  ...   

     Section 5 of 6 [I talk to a lot of different people at parties.]  \
881                                              (1, 5)                 
1197                                             (1, 3)                 

     Section 5 of 6 [I use difficult words.]  \
881                                   (5, 1)   
1197                                  (5, 5)   

     Section 5 of 6 [I worry about things.] Conscientiousness  \
881                                  (4, 1)                26   
1197                                 (4, 1)                26   

     Emotional Stability Intellect Agreeableness Extraversion    ID  \
881                   28        36            28           30   881   
1197                  26        28            22           40  1197   

       Department  
881          Data  
1197  Copywriting  

[2 rows x 57 columns]

### Print the ID numbers and departments of these applicants to the screen

In [54]:
individuals=score[['ID','Department']]

In [55]:
print(individuals)

        ID   Department
881    881         Data
1197  1197  Copywriting


###  Assign these applicants the tag “high risk” in a new column. All other applicants get the tag “low risk”

In [56]:
merge['Risk']=np.where((merge['Conscientiousness']<30) & (merge['Agreeableness']<30) & (merge['Emotional Stability'] <30),'high risk','low risk')

In [57]:
merge

Section 5 of 6 [I am always prepared.]  \
0                                    (3, 5)   
1                                    (3, 5)   
2                                    (3, 5)   
3                                    (3, 5)   
4                                    (3, 3)   
...                                     ...   
1550                                 (3, 5)   
1551                                 (3, 3)   
1552                                 (3, 5)   
1553                                 (3, 5)   
1554                                 (3, 3)   

     Section 5 of 6 [I am easily disturbed.]  \
0                                     (4, 5)   
1                                     (4, 5)   
2                                     (4, 3)   
3                                     (4, 5)   
4                                     (4, 5)   
...                                      ...   
1550                                  (4, 5)   
1551                                  (4, 5)   
1552                                  (4, 3)   
1553                                  (4, 5)   
1554                                  (4, 5)   

     Section 5 of 6 [I am exacting (demanding) in my work.]  \
0                                                (3, 5)       
1                                                (3, 5)       
2                                                (3, 3)       
3                                                (3, 3)       
4                                                (3, 3)       
...                                                 ...       
1550                                             (3, 1)       
1551                                             (3, 5)       
1552                                             (3, 5)       
1553                                             (3, 5)       
1554                                             (3, 1)       

     Section 5 of 6 [I am full of ideas.]  \
0                                  (5, 5)   
1                                  (5, 5)   
2                                  (5, 5)   
3                                  (5, 5)   
4                                  (5, 3)   
...                                   ...   
1550                               (5, 5)   
1551                               (5, 3)   
1552                               (5, 5)   
1553                               (5, 5)   
1554                               (5, 5)   

     Section 5 of 6 [I am interested in people.]  \
0                                         (2, 3)   
1                                         (2, 5)   
2                                         (2, 5)   
3                                         (2, 5)   
4                                         (2, 3)   
...                                          ...   
1550                                      (2, 5)   
1551                                      (2, 5)   
1552                                      (2, 5)   
1553                                      (2, 5)   
1554                                      (2, 5)   

     Section 5 of 6 [I am not interested in abstract ideas.]  \
0                                                (5, 3)        
1                                                (5, 3)        
2                                                (5, 5)        
3                                                (5, 3)        
4                                                (5, 3)        
...                                                 ...        
1550                                             (5, 5)        
1551                                             (5, 3)        
1552                                             (5, 5)        
1553                                             (5, 5)        
1554                                             (5, 3)        

     Section 5 of 6 [I am not interested in other people's problems.]  \
0                                                (2, 3)                 
1                        

### The number of low and high risk applicants within each department

In [58]:
counts=pd.crosstab(merge.Risk,merge.Department)
Applicants_totals=pd.DataFrame(counts)
Applicants_totals

Department  Copywriting  Data  Design  Strategy  Web Dev  Web dev
Risk                                                             
high risk             1     1       0         0        0        0
low risk            325   328     120       449      170      161